refactor(router): change type of `pathMatch`

WLun001 opened the pull request:
Union type has better IDE support, and avoid typos

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented:
This is not actually a docs change. It is an update to the public API, narrowing the type from `string` to a union of two string literals. Therefore the earliest that this could land is probably 9.1.0.

petebacondarwin commented:
Moreover, this appears to have broken the Angular build. See the various CI failures.

WLun001 commented:
Removed the test case
https://github.com/angular/angular/blob/master/packages/router/test/config.spec.ts#L126
It's not required to check anymore with string literal.

atscott reviewed (changes requested):
@WLun001 Could you please squash these commits into one and add a footer to indicate this is a breaking change (https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-footer)?

WLun001 committed:
refactor(router): improve type for Route.pathMatch

BREAKING CHANGE: strict typed for Route.pathMatch

restricted allowable values to 'prefix' or 'full'

jelbourn commented:
@atscott how likely is it that someone would read this setting from the `Route` config? It seems to me that you'd only really have reason to write it.

petebacondarwin commented:
@atscott - I have rebased and updated the commit message here. I think it is now in a position to merge before 12.0.0. Can you take a look?

atscott commented:
@petebacondarwin - this is a breaking change and since we’re in feature freeze, we can’t merge it anymore. Regardless, it’s a big breaking change and we need to figure out what the migration path would be.

atscott commented:
Closing since this cannot be merged as-is. The implementation LGTM but the PR needs an accompanying migration in order to be merged. I added the issue to v14 candidates to be reviewed when planning v14 migrations.
